File: python/graphs/bellman_ford.py
Cell = tuple[int, int]


def bellman_ford(
    width: int,
    height: int,
    walls: set[Cell],
    costs: dict[Cell, int],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    def neighbors(cell: Cell):
        row, col = cell
        for next_cell in (
            (row - 1, col),
            (row, col + 1),
            (row + 1, col),
            (row, col - 1),
        ):
            next_row, next_col = next_cell
            if 0 <= next_row < height and 0 <= next_col < width and next_cell not in walls:
                yield next_cell

    open_cells = [
        (row, col)
        for row in range(height)
        for col in range(width)
        if (row, col) not in walls
    ]
    distance = {start: 0}
    parent: dict[Cell, Cell] = {}

    for _ in range(len(open_cells) - 1):
        changed = False

        for cell in open_cells:
            if cell not in distance:
                continue

            for neighbor in neighbors(cell):
                candidate = distance[cell] + costs.get(neighbor, 1)
                if candidate >= distance.get(neighbor, float("inf")):
                    continue

                distance[neighbor] = candidate
                parent[neighbor] = cell
                changed = True

        if not changed:
            break

    for cell in open_cells:
        if cell not in distance:
            continue
        for neighbor in neighbors(cell):
            if distance[cell] + costs.get(neighbor, 1) < distance.get(neighbor, float("inf")):
                raise ValueError("reachable negative cycle")

    if target not in distance:
        return []

    path = [target]
    while path[-1] != start:
        path.append(parent[path[-1]])
    path.reverse()
    return path

File: java/graphs/BellmanFord.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BellmanFord {
    record Cell(int row, int col) {}

    public static List<Cell> bellmanFord(
            boolean[][] walls,
            int[][] costs,
            Cell start,
            Cell target
    ) {
        int height = walls.length;
        int width = walls[0].length;
        int total = width * height;
        int[][] distance = new int[height][width];
        Cell[][] parent = new Cell[height][width];

        for (int row = 0; row < height; row++) {
            java.util.Arrays.fill(distance[row], Integer.MAX_VALUE);
        }
        distance[start.row()][start.col()] = 0;

        for (int pass = 1; pass < total; pass++) {
            boolean changed = false;

            for (int row = 0; row < height; row++) {
                for (int col = 0; col < width; col++) {
                    if (walls[row][col] || distance[row][col] == Integer.MAX_VALUE) {
                        continue;
                    }

                    int[][] moves = {
                        {-1, 0},
                        {0, 1},
                        {1, 0},
                        {0, -1}
                    };

                    for (int[] move : moves) {
                        int nextRow = row + move[0];
                        int nextCol = col + move[1];
                        if (nextRow < 0 || nextRow >= height || nextCol < 0 || nextCol >= width
                                || walls[nextRow][nextCol]) {
                            continue;
                        }

                        int candidate = distance[row][col] + costs[nextRow][nextCol];
                        if (candidate < distance[nextRow][nextCol]) {
                            distance[nextRow][nextCol] = candidate;
                            parent[nextRow][nextCol] = new Cell(row, col);
                            changed = true;
                        }
                    }
                }
            }

            if (!changed) {
                break;
            }
        }

        for (int row = 0; row < height; row++) {
            for (int col = 0; col < width; col++) {
                if (walls[row][col] || distance[row][col] == Integer.MAX_VALUE) {
                    continue;
                }

                int[][] moves = {
                    {-1, 0},
                    {0, 1},
                    {1, 0},
                    {0, -1}
                };

                for (int[] move : moves) {
                    int nextRow = row + move[0];
                    int nextCol = col + move[1];
                    if (nextRow >= 0 && nextRow < height && nextCol >= 0 && nextCol < width
                            && !walls[nextRow][nextCol]
                            && distance[row][col] + costs[nextRow][nextCol] < distance[nextRow][nextCol]) {
                        throw new IllegalStateException("reachable negative cycle");
                    }
                }
            }
        }

        if (distance[target.row()][target.col()] == Integer.MAX_VALUE) {
            return List.of();
        }

        List<Cell> path = new ArrayList<>();
        for (Cell current = target; current != null; current = parent[current.row()][current.col()]) {
            path.add(current);
            if (current.equals(start)) {
                break;
            }
        }
        Collections.reverse(path);
        return path;
    }
}
